net/captivedetection/captivedetection.go

```diff
@@ -0,0 +1,189 @@
+func (d *Detector) DetectCaptivePortal(ifst *netmon.State, derpMap *tailcfg.DERPMap, preferredDERPRegionID int, logf logger.Logf, goos string) (found bool, err error) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	if logf == nil {
+		logf = log.Printf
+	}
+
+	if !ifst.AnyInterfaceUp() {
+		logf("DetectCaptivePortal: no interfaces up, returning false")
+		return false, nil
+	}
+
+	// Here we try detecting a captive portal using *all* available interfaces on the system
+	// that have a IPv4 address. We consider to have found a captive portal when any interface
+	// reports one may exists. This is necessary because most systems have multiple interfaces,
+	// and most importantly on macOS no default route interface is set until the user has accepted
+	// the captive portal alert thrown by the system. If no default route interface is known,
+	// we need to try with anything that might remotely resemble a Wi-Fi interface.
+	for ifName, i := range ifst.Interface {
+		if !i.IsUp() || i.IsLoopback() || interfaceNameDoesNotNeedCaptiveDetection(ifName, goos) {
+			continue
+		}
+		addrs, err := i.Addrs()
+		if err != nil {
+			logf("DetectCaptivePortal: failed to get addresses for interface %s: %v", ifName, err)
+			continue
+		}
+		if len(addrs) == 0 {
+			continue
+		}
+		logf("attempting to do captive portal detection on interface %s", ifName)
+		res := d.detectOnInterface(derpMap, i.Index, preferredDERPRegionID, logf, goos)
+		if res {
+			logf("DetectCaptivePortal(found=true,ifName=%s)", found, ifName)
+			return true, nil
+		}
+	}
+
+	logf("DetectCaptivePortal(found=false)")
+	return false, err
+}
+
+func interfaceNameDoesNotNeedCaptiveDetection(ifName string, goos string) bool {
+	ifName = strings.ToLower(ifName)
+	excludedPrefixes := []string{"tailscale", "tun", "tap", "docker", "kube", "wg"}
+	if goos == "windows" {
+		excludedPrefixes = append(excludedPrefixes, "loopback", "tunnel", "ppp", "isatap", "teredo", "6to4")
+	} else if goos == "darwin" || goos == "ios" {
+		excludedPrefixes = append(excludedPrefixes, "awdl", "bridge", "ap", "utun", "tap", "llw", "anpi", "lo", "stf", "gif", "xhc")
+	}
+	for _, prefix := range excludedPrefixes {
+		if strings.HasPrefix(ifName, prefix) {
+			return true
+		}
+	}
+	return false
+}
+
+// detectOnInterface reports whether or not we think the system is behind a
+// captive portal, detected by making a request to a URL that we know should
+// return a "204 No Content" response and checking if that's what we get.
+//
+// The boolean return is whether we think we have a captive portal.
+func (d *Detector) detectOnInterface(derpMap *tailcfg.DERPMap, ifIndex int, preferredDERPRegionID int, logf logger.Logf, goos string) bool {
+	defer d.httpClient.CloseIdleConnections()
+
+	endpoints := availableEndpoints(derpMap, preferredDERPRegionID, logf, goos)
+	logf("%d available captive portal detection endpoints: %v", len(endpoints), endpoints)
```

Comment:
It looks like this doesn't change based on the interface we're checking (since we don't use `ifIndex` here). You could pull this out to the top-level function and then just pass `endpoints` around, instead of `derpMap`, `preferredDERPRegionID`, and `goos`?

Reply:
Done